### PR TITLE
Basic Avocado tests structure

### PR DESCRIPTION
**Pull request by apahim**
Introduce the basic structure to run Avocado tests.
No extra dependencies are required other than the Avocado core package itself.

### CODE REVIEW

**Review by adereis — COMMENTED** [review]
When adding references to a Red Hat BZ:

Use "RHBZ ID: URL" for Red Hat Bugzillas (it's not an official convention, but you'll see some of this in the upstream git log)

If the summary (BZ Title) is relevant and descritive, add it there somewhere. Otherwise, briefly describe what the problem is in the changelog.

File: tests/avocado/test_ovmf_with_240_vcpus.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+from avocado import Test
+from base import VM
+
+
+class TestOvmfVcpus(Test):
+    """
+    :avocado: tags=ovmf
+    """
+
+    def setUp(self):
```
Comment (adereis) [inline tests/avocado/test_ovmf_with_240_vcpus.py:14]: Here we see a clear need for some sort of dependency support. This test has several environment dependencies and should be skipped if these dependencies are not met.

I wrote something about it in the avocado-devel discussion the other day, in the context of the thread "RFC: Guidelines for categorizing tests": http://www.mail-archive.com/avocado-devel@redhat.com/msg00643.html and http://www.mail-archive.com/avocado-devel@redhat.com/msg00645.html

File: tests/avocado/test_ovmf_with_240_vcpus.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        args.extend(['-drive', 'file=%s' % ovmf_vars])
+
+        self.vm = VM(args)
+
+    def test_run_vm(self):
+        """
+        According to the BZ1447027, the issue is: Guest cannot boot with
+        240 or above vcpus when using ovmf.
+
+        :avocado: tags=BZ1447027
```
Comment (adereis) [inline tests/avocado/test_ovmf_with_240_vcpus.py:35]: Please use RHBZ when referring to the Red Hat Bugzilla in the upstream qemu context.

File: tests/avocado/test_info_memdev_host_nodes.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        cmd = 'device_add pc-dimm,id=dimm1,memdev=mem1'
+        res = self.vm.qmp('human-monitor-command', command_line=cmd)
+        self.assertEqual('', res['return'])
+
+        cmd = 'info memdev'
+        res = self.vm.qmp('human-monitor-command', command_line=cmd)
+        self.assertIn('host nodes: \r', res['return'])
+
+    def tearDown(self):
+        self.vm.shutdown()
```
Comment (adereis) [inline tests/avocado/test_info_memdev_host_nodes.py:36]: Please add:

```
if __name__ == "__main__":
    avocado.main()
```

This way users can run the test individually without invoking Avocado.

File: tests/avocado/README.rst [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        def tearDown(self):
+            self.vm.shutdown()
+
+To execute your test, run::
+
+    avocado run tess_my_test.py
+
+To execute all tests, run::
+
+    avocado run .
```
Comment (adereis) [inline tests/avocado/README.rst:39]: What about ending with something like:

> see `avocado run --help` and `man avocado` for several other options, such as `--filter-by-tags`, `--show-job-log`, `--failfast`, etc.

**Review by adereis — COMMENTED** [review]
This is a great start. Let me give a few more suggestions (which you might have though of already but had no time to implement):

 * Provide a default avocado configuration file in this directory with features which should be interesting to most upstream developers (things like `test_dir`, `logs_dir`, `sysinfo_collect=False`, etc). Will avocado read a `.avocado` configuration file if it's present in the current directory? If not, perhaps it should.

* Several tests need (or will need) parametrization. For example, the location of a file (such as the ovmf binaries), the name/location of the QEMU binary, IP Addresses, ports, etc. To allow good scalability, these should never be hardcoded inside the test, but instead be configured somewhere and requested inside the source code via the varianter test-API so users can re-run the same tests with a different environment file and even extend the test using the multiplexer (but tests would work without it by default). Do we have support for providing the parameters (the "default variant") without using the full multiplexer? In just a few words, we should have a way to provide the "default variant" via multiple hierarchical sources, similar to the way we read our configuration (global environment file, local environment file, test source code, `--test-env` and, finally, the override of other variants via the multiplexer (Varianter)). I would implement it as a simple .ini file (and/or multiple of them). Then in this directory there would be a qemu-tests.ini with some defaults, tests would be free to have their own .ini as well and users could implement multiple .ini files (fedora.ini, ubuntu.ini, s390.ini, etc). Again, to avoid confusion: the .ini file contains the "default variant". It shares the Test-API with the multiplexer/varianter.

* I think using qemu.py from `qemu.git/scripts` is a good start, but before investing too much time writing tests, have you considered (and discarded) the idea of implementing a full avocado plugin for qemu? For example, one that introduces new command-line options and support for advanced features like taking a snapshot of the image before running a test, templates for QEMU command-line options, new options to the config file, library of images, etc? This plugin could live in qemu.git.

File: tests/avocado/test_ovmf_with_240_vcpus.py (hunk 2 above)
Comment (adereis) [reply tests/avocado/test_ovmf_with_240_vcpus.py:35]: To the upstream developers, way more important the Red Hat Bugzilla ID is a reference to a commit that fixed the Bug. I mean, adding a RHBZ reference is welcome, but if you can find the commit ID with a simple search, you'll add considerable value to your test and changelog.

File: tests/avocado/test_ovmf_with_240_vcpus.py (hunk 1 above)
Comment (apahim) [reply tests/avocado/test_ovmf_with_240_vcpus.py:14]: Agreed.

File: tests/avocado/test_ovmf_with_240_vcpus.py (hunk 2 above)
Comment (apahim) [reply tests/avocado/test_ovmf_with_240_vcpus.py:35]: Ok

File: tests/avocado/test_info_memdev_host_nodes.py (hunk 3 above)
Comment (apahim) [reply tests/avocado/test_info_memdev_host_nodes.py:36]: Sure

File: tests/avocado/README.rst (hunk 4 above)
Comment (apahim) [reply tests/avocado/README.rst:39]: Sure, ok

File: tests/avocado/test_ovmf_with_240_vcpus.py (hunk 1 above)
Comment (adereis) [reply tests/avocado/test_ovmf_with_240_vcpus.py:14]: If we still don't have dependency-tags or something like `checkDeps()` (that would run before `setUp()`), then we could use decorators, IIRC.

My point is that a test that runs `shutil.copy('/usr/share/edk2/ovmf/OVMF_VARS.fd' ...` will FAIL with errors most of the time in the context of upstream.

Comment (apahim) [reply tests/avocado/test_ovmf_with_240_vcpus.py:14]: Yes, the best approach here is indeed to use the skip decorators.

**Comment by apahim** [comment]
> Provide a default avocado configuration file in this directory with features which should be interesting to most upstream developers (things like test_dir, logs_dir, sysinfo_collect=False, etc). Will avocado read a .avocado configuration file if it's present in the current directory? If not, perhaps it should.

https://trello.com/c/8qc7wwZL/1075-support-avocado-configuration-file-if-its-present-in-the-current-directory

> Several tests need (or will need) parametrization. For example, the location of a file (such as the ovmf binaries), the name/location of the QEMU binary, IP Addresses, ports, etc. To allow good scalability, these should never be hardcoded inside the test, but instead be configured somewhere and requested inside the source code via the varianter test-API so users can re-run the same tests with a different environment file and even extend the test using the multiplexer (but tests would work without it by default). Do we have support for providing the parameters (the "default variant") without using the full multiplexer? In just a few words, we should have a way to provide the "default variant" via multiple hierarchical sources, similar to the way we read our configuration (global environment file, local environment file, test source code, --test-env and, finally, the override of other variants via the multiplexer (Varianter)). I would implement it as a simple .ini file (and/or multiple of them). Then in this directory there would be a qemu-tests.ini with some defaults, tests would be free to have their own .ini as well and users could implement multiple .ini files (fedora.ini, ubuntu.ini, s390.ini, etc). Again, to avoid confusion: the .ini file contains the "default variant". It shares the Test-API with the multiplexer/varianter.

Well, I was thinking in provide the parameters using full multiplex, keeping a `parameters.yaml` file in the base directory and additional yaml files to provide specific parameters and variants for each test (and override the base parameters, if needed) could be present in a test directory, like:
```
[~/git/qemu/tests/avocado]$ cat parameters.yaml 
key1: val1
key2: val2
```
```
[~/git/qemu/tests/avocado]$ cat test1.py.data/parameters.yaml 
key2: val_other
key3: val3
```
```
[~/git/qemu/tests/avocado]$ cat test1.py
import avocado
import logging

LOG = logging.getLogger('mylog')

class MyTest(avocado.Test):
    def test(self):
        LOG.info(self.params.get('key1', default='default1'))
        LOG.info(self.params.get('key2', default='default2'))
        LOG.info(self.params.get('key3', default='default3'))
```
```
[~/git/qemu/tests/avocado]$ avocado --show mylog run test1.py -m parameters.yaml test1.py.data/parameters.yaml
mylog: val1
mylog: val_other
mylog: val3
```
Do you think that's enough?

 > I think using qemu.py from qemu.git/scripts is a good start, but before investing too much time writing tests, have you considered (and discarded) the idea of implementing a full avocado plugin for qemu? For example, one that introduces new command-line options and support for advanced features like taking a snapshot of the image before running a test, templates for QEMU command-line options, new options to the config file, library of images, etc?

I have considered, yes, but that's avocado-virt. We can turn it into an optional plugin, package it along the others plugins and so on. That would make the dependency easier to satisfy. I like the idea of using avocado-virt, but not sure if I am the only one. Am I?

**Comment by apahim** [comment]
Closed in favour of #2

**Comment by adereis** [comment]
On Wed, Jun 28, 2017 at 11:05:11PM +0000, Amador Pahim wrote:
> > Provide a default avocado configuration file in this directory with features which should be interesting to most upstream developers (things like test_dir, logs_dir, sysinfo_collect=False, etc). Will avocado read a .avocado configuration file if it's present in the current directory? If not, perhaps it should.
> 
> https://trello.com/c/8qc7wwZL/1075-support-avocado-configuration-file-if-its-present-in-the-current-directory
> 
> > Several tests need (or will need) parametrization. For example, the location of a file (such as the ovmf binaries), the name/location of the QEMU binary, IP Addresses, ports, etc. To allow good scalability, these should never be hardcoded inside the test, but instead be configured somewhere and requested inside the source code via the varianter test-API so users can re-run the same tests with a different environment file and even extend the test using the multiplexer (but tests would work without it by default). Do we have support for providing the parameters (the "default variant") without using the full multiplexer? In just a few words, we should have a way to provide the "default variant" via multiple hierarchical sources, similar to the way we read our configuration (global environment file, local environment file, test source code, --test-env and, finally, the override of other variants via the multiplexer (Varianter)). I would implement it as a simple .ini file (and/or multiple of them). Then in this directory there would be a qemu-tests.ini with some defaults, tests would be free to have their own .ini as well and users could implement multiple .ini files (fedora.ini, ubuntu.ini, s390.ini, etc). Again, to avoid confusion: the .ini file contains the "default variant". It shares the Test-API with the multiplexer/varianter.
> 
> Well, I was thinking in provide the parameters using full
> multiplex, keeping a `parameters.yaml` file in the base
> directory and additional yaml files to provide specific
> parameters and variants for each test (and override the base
> parameters, if needed) could be present in a test directory,
> like:

I think using the multiplexer is overkill for this kind of thing
but...

> ```
> [~/git/qemu/tests/avocado]$ cat parameters.yaml 
> key1: val1
> key2: val2
> ```
> ```
> [~/git/qemu/tests/avocado]$ cat test1.py.data/parameters.yaml 
> key2: val_other
> key3: val3
> ```

...looking at these simple yaml files, they're not that different
from a .ini file, so it might be enough.

> ```
> [~/git/qemu/tests/avocado]$ cat test1.py
> import avocado
> import logging
> 
> LOG = logging.getLogger('mylog')
> 
> class MyTest(avocado.Test):
>     def test(self):
>         LOG.info(self.params.get('key1', default='default1'))
>         LOG.info(self.params.get('key2', default='default2'))
>         LOG.info(self.params.get('key3', default='default3'))
> ```
> ```
> [~/git/qemu/tests/avocado]$ avocado --show mylog run test1.py -m parameters.yaml test1.py.data/parameters.yaml
> mylog: val1
> mylog: val_other
> mylog: val3
> ```
> Do you think that's enough?

Looks good, indeed.

I usually don't think of the multiplexer/varianter as something
simple for basic test configuration. I see it as an optional and
much more complex component and a much harder to understand
feature.  But without a yaml tree it can be seen as a simple
mechanism to provide parameters, which is exactly what I want.

> 
>  > I think using qemu.py from qemu.git/scripts is a good start,
>  > but before investing too much time writing tests, have you
>  > considered (and discarded) the idea of implementing a full
>  > avocado plugin for qemu? For example, one that introduces
>  > new command-line options and support for advanced features
>  > like taking a snapshot of the image before running a test,
>  > templates for QEMU command-line options, new options to the
>  > config file, library of images, etc?
> 
> I have considered, yes, but that's avocado-virt. We can turn it
> into an optional plugin, package it along the others plugins
> and so on. That would make the dependency easier to satisfy. I
> like the idea of using avocado-virt, but not sure if I am the
> only one. Am I?
> 

I was thinking of something like `avocado-qemu` (re-using some of
what's in avocado-virt), but maintained inside `qemu.git`.

-- 
Ademar Reis
Red Hat

^[:wq!

**Pull request closed without merging**